torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+    # inputs given as arguments and the same Tensors automatically captured by the user function.
+    # Check this issue for more details on how that can happen: https://github.com/pytorch/pytorch/issues/32576
+    res = []
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
+        else:
+            res.append(inp.detach().requires_grad_(need_graph))
+    return tuple(res)
+
+
+def _grad_postprocess(inputs, create_graph):
+    # Postprocess the generated Tensors to avoid returning Tensors with history when the user did not
+    # request it.
+    if torch.is_tensor(inputs[0]):
+        if not create_graph:
+            return tuple(inp.detach() for inp in inputs)
+        else:
+            return inputs
+    else:
+        return tuple(_grad_postprocess(inp, create_graph) for inp in inputs)
+
+def _validate_v(v, other):
+    # This assumes that other is the correct shape, and v should match
+    # Both are assumed to be tuples of Tensors
+    if len(other) != len(v):
+        raise RuntimeError("v is a tuple of invalid length: should be {} but got {}.".format(len(other), len(v)))
+
+    for idx, (el_v, el_other) in enumerate(zip(v, other)):
+        if el_v.size() != el_other.size():
+            raise RuntimeError("Entry {} in v has invalid size: should be {} but got {}.".format(
+                               idx, el_other.size(), el_v.size()))
+
+def _check_requires_grad(inputs, input_type, strict):
+    # Used to make all the necessary checks to raise nice errors in strict mode.
+    if not strict:
+        return
+
+    if input_type not in ["outputs", "grad_inputs", "jacobian", "hessian"]:
+        raise RuntimeError("Invalid input_type to _check_requires_grad")
+    for i, inp in enumerate(inputs):
+        if inp is None:
+            # This can only be reached for grad_inputs.
+            raise RuntimeError("The output of the user-provided function is independent of input {}."
+                               " This is not allowed in strict mode.".format(i))
+        if not inp.requires_grad:
+            if input_type == "hessian":
+                raise RuntimeError("The hessian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is thrice differentiable and that"
+                                   " the hessian depends on the inputs.".format(i))
+            elif input_type == "jacobian":
+                raise RuntimeError("The jacobian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is twice differentiable and that"
```

Comment:
maybe a dumb question, but why do I need my function to be twice differentiable to compute the jacobian?

Reply:
This is used when we compute the Jacobian as an intermediary result for a hessian. In that case, we need the Jacobian to require gradients.

Reply:
could you make the error message clearer, like "While computing the hessian, found that the Jacobian ..."

Reply:
Sure